File: server-shared/src/test/java/rikka/shizuku/server/ServerTestSupport.java
package rikka.shizuku.server;

import static org.mockito.Mockito.CALLS_REAL_METHODS;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.os.Bundle;
import android.os.IBinder;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.function.IntFunction;

import eu.darken.porter.core.CallerIdentity;
import eu.darken.porter.core.PorterCore;
import eu.darken.porter.core.ServerPolicy;
import moe.shizuku.server.IShizukuApplication;

/** The server pieces every suite in this module builds its subject out of. */
public final class ServerTestSupport {

    private ServerTestSupport() {
    }

    /** A policy that answers what a test set on it and remembers what it was told. */
    public static class TestPolicy implements ServerPolicy {

        public boolean callerPermission;
        public boolean managerPermission;

        public boolean confirmationShown;
        public int confirmationRequestCode;
        public ClientRecord confirmationRecord;
        public int confirmationUid;
        public int confirmationPid;
        public int confirmationUserId;

        public final List<CallerIdentity> attachingCallers = new ArrayList<>();
        public final List<String> attachingPackages = new ArrayList<>();
        public final List<ClientRecord> attachedRecords = new ArrayList<>();
        public final List<Boolean> attachedCreated = new ArrayList<>();
        public final List<Bundle> attachedReplies = new ArrayList<>();
        public final List<ClientRecord> boundRecords = new ArrayList<>();
        public final List<Boolean> boundCreated = new ArrayList<>();

        @Override
        public boolean checkCallerPermission(String func, CallerIdentity caller, ClientRecord record) {
            return callerPermission;
        }

        @Override
        public boolean checkCallerManagerPermission(String func, CallerIdentity caller) {
            return managerPermission;
        }

        @Override
        public void showPermissionConfirmation(
                int requestCode, ClientRecord record, CallerIdentity caller, int userId) {
            confirmationShown = true;
            confirmationRequestCode = requestCode;
            confirmationRecord = record;
            confirmationUid = caller.uid;
            confirmationPid = caller.pid;
            confirmationUserId = userId;
        }

        @Override
        public void onAttaching(CallerIdentity caller, String packageName) {
            attachingCallers.add(caller);
            attachingPackages.add(packageName);
        }

        @Override
        public void onAttached(ClientRecord record, boolean created, Bundle reply) {
            attachedRecords.add(record);
            attachedCreated.add(created);
            attachedReplies.add(reply);
        }

        @Override
        public void onBound(ClientRecord record, boolean created) {
            boundRecords.add(record);
            boundCreated.add(created);
        }
    }

    /** A concrete manager so the core carries a real object rather than a mock. */
    public static class TestUserServiceManager extends UserServiceManager {
        @Override
        public String getUserServiceStartCmd(
                UserServiceRecord record, String key, String token, String packageName,
                String classname, String processNameSuffix, int callingUid, boolean use32Bits, boolean debug) {
            return "exit 0";
        }
    }

    public static PorterCore<UserServiceManager, ClientManager<ConfigManager>, ConfigManager> newCore(
            ClientManager<ConfigManager> clientManager,
            UserServiceManager userServiceManager,
            ConfigManager configManager,
            ServerPolicy policy,
            IntFunction<List<String>> packagesForUid) {
        return new PorterCore<>(userServiceManager, clientManager, configManager, policy, packagesForUid);
    }

    /**
     * The six {@code IShizukuService} methods {@link Service} does not implement stay abstract;
     * {@code mock(TestService.class, CALLS_REAL_METHODS)} answers those with defaults.
     */
    public abstract static class TestService
            extends Service<UserServiceManager, ClientManager<ConfigManager>, ConfigManager> {

        public boolean callerPermission;
        public boolean managerPermission;

        public boolean confirmationShown;
        public int confirmationRequestCode;
        public ClientRecord confirmationRecord;
        public int confirmationUid;
        public int confirmationPid;
        public int confirmationUserId;

        public IShizukuApplication attachedApplication;
        public Bundle attachedArgs;

        @Override
        public UserServiceManager onCreateUserServiceManager() {
            return null;
        }

        @Override
        public ClientManager<ConfigManager> onCreateClientManager() {
            return null;
        }

        @Override
        public ConfigManager onCreateConfigManager() {
            return null;
        }

        @Override
        public boolean checkCallerPermission(String func, int callingUid, int callingPid, ClientRecord clientRecord) {
            return callerPermission;
        }

        @Override
        public boolean checkCallerManagerPermission(String func, int callingUid, int callingPid) {
            return managerPermission;
        }

        @Override
        public void showPermissionConfirmation(
                int requestCode, ClientRecord clientRecord, int callingUid, int callingPid, int userId) {
            confirmationShown = true;
            confirmationRequestCode = requestCode;
            confirmationRecord = clientRecord;
            confirmationUid = callingUid;
            confirmationPid = callingPid;
            confirmationUserId = userId;
        }

        @Override
        public void attachApplication(IShizukuApplication application, Bundle args) {
            attachedApplication = application;
            attachedArgs = args;
        }
    }

    /**
     * {@code Service()} loads the native porsh library, so the service under test is built without
     * running any constructor and the three manager fields are injected.
     */
    public static TestService newService(
            ClientManager<ConfigManager> clientManager, UserServiceManager userServiceManager, ConfigManager configManager) {
        TestService service = mock(TestService.class, CALLS_REAL_METHODS);
        inject(service, "clientManager", clientManager);
        inject(service, "userServiceManager", userServiceManager);
        inject(service, "configManager", configManager);
        return service;
    }

    static void inject(Object service, String name, Object value) {
        try {
            Field field = Service.class.getDeclaredField(name);
            field.setAccessible(true);
            field.set(service, value);
        } catch (ReflectiveOperationException e) {
            throw new AssertionError(e);
        }
    }

    /** An application binder death can be fired from, which {@code addClient} needs to link to. */
    public static IShizukuApplication application(IBinder binder) {
        IShizukuApplication application = mock(IShizukuApplication.class);
        when(application.asBinder()).thenReturn(binder);
        return application;
    }

    public static ConfigPackageEntry entry(boolean allowed, boolean denied) {
        return new ConfigPackageEntry() {
            @Override public boolean isAllowed() { return allowed; }
            @Override public boolean isDenied() { return denied; }
        };
    }
}

File: server-shared/src/test/java/rikka/shizuku/server/ServerTestSupport.kt
package rikka.shizuku.server

import android.os.Bundle
import android.os.IBinder
import eu.darken.porter.core.CallerIdentity
import eu.darken.porter.core.PorterCore
import eu.darken.porter.core.ServerPolicy
import java.util.function.IntFunction
import moe.shizuku.server.IShizukuApplication
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`

/** The server pieces every suite in this module builds its subject out of. */
object ServerTestSupport {

    /** A policy that answers what a test set on it and remembers what it was told. */
    open class TestPolicy : ServerPolicy {

        var callerPermission = false
        var managerPermission = false

        var confirmationShown = false
        var confirmationRequestCode = 0
        var confirmationRecord: ClientRecord? = null
        var confirmationUid = 0
        var confirmationPid = 0
        var confirmationUserId = 0

        val attachingCallers: MutableList<CallerIdentity> = ArrayList()
        val attachingPackages: MutableList<String> = ArrayList()
        val attachedRecords: MutableList<ClientRecord> = ArrayList()
        val attachedCreated: MutableList<Boolean> = ArrayList()
        val attachedReplies: MutableList<Bundle> = ArrayList()
        val boundRecords: MutableList<ClientRecord> = ArrayList()
        val boundCreated: MutableList<Boolean> = ArrayList()

        override fun checkCallerPermission(func: String, caller: CallerIdentity, record: ClientRecord?): Boolean =
            callerPermission

        override fun checkCallerManagerPermission(func: String, caller: CallerIdentity): Boolean =
            managerPermission

        override fun showPermissionConfirmation(requestCode: Int, record: ClientRecord, caller: CallerIdentity, userId: Int) {
            confirmationShown = true
            confirmationRequestCode = requestCode
            confirmationRecord = record
            confirmationUid = caller.uid
            confirmationPid = caller.pid
            confirmationUserId = userId
        }

        override fun onAttaching(caller: CallerIdentity, packageName: String) {
            attachingCallers.add(caller)
            attachingPackages.add(packageName)
        }

        override fun onAttached(record: ClientRecord, created: Boolean, reply: Bundle) {
            attachedRecords.add(record)
            attachedCreated.add(created)
            attachedReplies.add(reply)
        }

        override fun onBound(record: ClientRecord, created: Boolean) {
            boundRecords.add(record)
            boundCreated.add(created)
        }
    }

    /** A concrete manager so the core carries a real object rather than a mock. */
    open class TestUserServiceManager : UserServiceManager() {
        override fun getUserServiceStartCmd(
            record: UserServiceRecord,
            key: String,
            token: String,
            packageName: String,
            classname: String,
            processNameSuffix: String?,
            callingUid: Int,
            use32Bits: Boolean,
            debug: Boolean,
        ): String = "exit 0"
    }

    fun newCore(
        clientManager: ClientManager<ConfigManager>,
        userServiceManager: UserServiceManager,
        configManager: ConfigManager,
        policy: ServerPolicy,
        packagesForUid: IntFunction<List<String>>,
    ): PorterCore<UserServiceManager, ClientManager<ConfigManager>, ConfigManager> =
        PorterCore(userServiceManager, clientManager, configManager, policy, packagesForUid)

    /** An application binder death can be fired from, which `addClient` needs to link to. */
    fun application(binder: IBinder): IShizukuApplication {
        val application = mock(IShizukuApplication::class.java)
        `when`(application.asBinder()).thenReturn(binder)
        return application
    }

    fun entry(allowed: Boolean, denied: Boolean): ConfigPackageEntry = object : ConfigPackageEntry() {
        override fun isAllowed(): Boolean = allowed
        override fun isDenied(): Boolean = denied
    }
}
